include/bn256.h

```diff
@@ -0,0 +1,168 @@
+
+#include <boost/multiprecision/cpp_int.hpp>
+#include <utility>
+#include <array>
+#include <system_error>
+#include "span.h"
+
+using boost::multiprecision::int512_t;
+
+namespace bn256 {
+
+struct curve_point;
+struct twist_point;
+struct gfp12;
+
+// G1 is an abstract cyclic group. The zero value is suitable for use as the
+// output of an operation, but cannot be used as an input.
+class g1 {
+   uint64_t p_[4 * 4];
+
+ public:
+   g1() = default;
+   explicit g1(const curve_point&);
+
+   static g1 curve_gen;
+
+   curve_point& p();
+   const curve_point& p() const;
+
+   static g1 scalar_base_mult(const int512_t& k) noexcept;
+
+   g1 scalar_mult(const int512_t& k) const noexcept;
+
+   g1 add(const g1& b) const noexcept;
+
+   g1 neg();
+
+   void                    marshal(std::span<uint8_t, 64> out) const noexcept;
+   std::array<uint8_t, 64> marshal() const noexcept {
+      std::array<uint8_t, 64> result;
+      marshal(result);
+      return result;
+   }
+
+   [[nodiscard]] std::error_code unmarshal(std::span<const uint8_t, 64> in) noexcept;
+
+   bool operator==(const g1& rhs) const noexcept { return memcmp(p_, rhs.p_, sizeof(*this)) == 0; }
+   bool operator!=(const g1& rhs) const noexcept { return !(*this == rhs); }
+
+   std::string string() const;
+};
+
+inline std::ostream& operator<<(std::ostream& os, const g1& v) { return os << v.string(); }
+
+// g2 is an abstract cyclic group. The zero value is suitable for use as the
+// output of an operation, but cannot be used as an input.
+class g2 {
+   uint64_t p_[4 * 2 * 4];
+
+ public:
+   g2() = default;
+   explicit g2(const twist_point&);
+
+   static g2 twist_gen;
+
+   twist_point& p();
+   const twist_point& p() const;
+
+   static g2 scalar_base_mult(const int512_t& k) noexcept;
+
+   g2 scalar_mult(const int512_t& k) const noexcept;
+
+   g2 add(const g2& b) const noexcept;
+
+   g2 neg() const noexcept;
+
+   void                     marshal(std::span<uint8_t, 128> out) const noexcept;
+   std::array<uint8_t, 128> marshal() const noexcept {
+      std::array<uint8_t, 128> result;
+      marshal(result);
+      return result;
+   }
+
+   [[nodiscard]] std::error_code unmarshal(std::span<const uint8_t, 128> m) noexcept;
+
+   bool operator==(const g2& rhs) const noexcept { return memcmp(p_, rhs.p_, sizeof(*this)) == 0; }
+   bool operator!=(const g2& rhs) const noexcept { return !(*this == rhs); }
+
+   std::string string() const;
+};
+
+inline std::ostream& operator<<(std::ostream& os, const g2& v) { return os << v.string(); }
+
+// GT is an abstract cyclic group. The zero value is suitable for use as the
+// output of an operation, but cannot be used as an input.
+class gt {
+
+   uint64_t p_[4 * 2 * 3 * 2];
+
+ public:
+   gt() = default;
+   explicit gt(const gfp12&);
+
+   gfp12& p();
+   const gfp12& p() const;
+
+   gt scalar_mult(const int512_t& k) const noexcept;
+
+   gt add(const gt& b) const noexcept;
```

Comment:
This might be too much to ask. Can `add`, `neg`, `sub` be changed to operators + and - like == and != below?

Reply:
We are given the instruction that the code should be as close as to the origin golang code as much as possible; change the code to use operator overloading would make it harder to match to origin code.

Reply:
Yes, that's initial idea to keep as close as possible to the original code. Style changes might be OK. Either way is fine. Don't make any further changes.